src/bank.rs

```diff
@@ -431,15 +447,20 @@ impl Bank {
 
     /// Process a Witness Signature. Any payment plans waiting on this signature
     /// will progress one step.
-    fn apply_signature(&self, from: PublicKey, tx_sig: Signature) -> Result<()> {
+    fn apply_signature(
+        &self,
+        from: PublicKey,
+        contract_id: Signature,
+        balances: &mut HashMap<PublicKey, i64>,
```

Comment:
Added balances here to fix deadlock in previous code where we try to grab the write lock on the balances table while still holding the write lock from apply_credits(). Same fix was made for apply_timestamp().